aeppl/scan.py

```diff
@@ -0,0 +1,332 @@
+    if output_scan_args.n_steps is None or replacing:
+        output_scan_args.n_steps = n_steps
+
+    output_scan_args.outer_in_seqs += [v[:n_steps] for v in var_slices]
+
+    if not replacing or add_nit_sot:
+        output_scan_args.outer_in_nit_sot += [n_steps]
+
+    if inner_out_fn:
+        output_scan_args.inner_out_nit_sot += [
+            inner_out_fn(
+                input_scan_args, old_inner_out_var, inner_in_var, output_scan_args
+            )
+        ]
+
+    return output_scan_args, new_outer_input_var
+
+
+def get_random_outer_outputs(scan_args: ScanArgs) -> List[TensorVariable]:
+    """Get the `MeasurableVariable` outputs of a `Scan` (well, its `ScanArgs`)."""
+    rv_vars = []
+    for n, oo in enumerate(scan_args.outer_outputs):
+        oo_info = scan_args.find_among_fields(oo)
+        io_type = oo_info.name[(oo_info.name.index("_", 6) + 1) :]
+        inner_out_type = "inner_out_{}".format(io_type)
+        io_var = getattr(scan_args, inner_out_type)[oo_info.index]
+        if io_var.owner and isinstance(io_var.owner.op, MeasurableVariable):
+            rv_vars.append((n, oo, io_var))
+    return rv_vars
+
+
+def construct_scan(scan_args: ScanArgs) -> List[TensorVariable]:
+    scan_op = Scan(scan_args.inner_inputs, scan_args.inner_outputs, scan_args.info)
+    scan_out = scan_op(*scan_args.outer_inputs)
+
+    if not isinstance(scan_out, list):
+        scan_out = [scan_out]
+
+    return scan_out
+
+
+def create_inner_out_logp(
+    input_scan_args: ScanArgs,
+    old_inner_out_var: TensorVariable,
+    new_inner_in_var: TensorVariable,
+    output_scan_args: ScanArgs,
+) -> TensorVariable:
+    """Create a log-likelihood inner-output for a `Scan`."""
+    logp = _logprob(
+        old_inner_out_var.owner.op, new_inner_in_var, *old_inner_out_var.owner.inputs
+    )
+    if new_inner_in_var.name:
+        logp.name = "logp({})".format(new_inner_in_var.name)
+    return logp
+
+
+@_logprob.register(MeasurableScan)
+def logprob_ScanRV(op, value, *inputs, name=None, **kwargs):
+
+    # TODO: Use `value`?
+    rv_outer_outs = get_random_outer_outputs(op.scan_args)
+
+    for var_idx, var, io_var in rv_outer_outs:
+        scan_args, new_oi_var = convert_outer_out_to_in(
+            op.scan_args,
+            var,
+            inner_out_fn=create_inner_out_logp,
+            output_scan_args=op.scan_args,
+        )
+
+    logp_scan_out = construct_scan(scan_args)
+
+    return logp_scan_out
+
+
+@local_optimizer([Scan])
+def add_opts_to_inner_graphs(fgraph, node):
+    """Update the `Mode`(s) used to compile the inner-graph of a `Scan` `Op`.
+
+    This is how we add `rv_sinking_db` and any other log-probability rewrites
+    to the "body" (i.e. inner-graph) of a `Scan` loop.
+    """
+
+    if not isinstance(node.op, Scan):
+        return None
+
+    op = Scan(node.op.inputs, node.op.outputs, node.op.info)
+    new_node = op.make_node(*node.inputs)
+    new_node.op.mode = new_node.op.mode.register([rv_sinking_db])
```

Comment:
Isn't the mode itself `None` over here.

Reply:
I just threw that in there to be safe, and it might not even be useful.  There are two `Mode` values attached to the `Op`, the I don't think `mode` is the one that's directly used to compile the inner-graph; however, it may be used in some other way.